### ISSUE TITLE
Bug: TypeScript: Passing a Component Constructor as prop does not adhere to prop's type

### ISSUE THREAD

**Issue by NicoleRauch** [facebook/react#19517]
When passing a component constructor as prop, the component constructor type is not required to match the prop's type, which can lead to issues further down the execution path.

Here is an example:
https://codepen.io/nicolerauch/pen/QWNLENx?editors=1010

1. Let's say we have a component `EntryFunctional`/`EntryClass` (with props `EntryProps`) that we want to pass to the component `List`. For this, we equip `List` with a prop called `EntryComp` and give it the type `((props: EntryProps) => ReactElement) | (new (p: EntryProps) => Component<EntryProps>)` (so that it can handle both functional and class components).
When passing our `EntryFunctional` and `EntryClass` components (line 58 and 59), all is well.

2. If we have a component that accepts fewer props than `EntryProps` (in our case, no props at all), this should be accepted as well, according to https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-with-fewer-parameters-assignable-to-functions-that-take-more-parameters: If the component gets more props than it needs, that's fair enough.
In the example I provided two components `FewerPropsFunctional` and `FewerPropsClass`, but only the functional component is accepted by `List`, not the class component (lines 60 and 61).
I think this is **bug no. 1**: The class component should be accepted as well, because it requires fewer props than it is given.

3. If we have a component that requires more props than `EntryProps`, this should not be accepted because at runtime the component will not get all the data it needs.
In the example I provided two components `MorePropsFunctional` and `MorePropsClass`. Those are both accepted when passed to `List`. As you can see in the output, the `more` prop is `undefined` although its prop type is `number`. I think this is **bug no. 2**: Both components should not be accepted (i.e. these lines should lead to compile errors), as they do not match the type of the prop and as this leads to breaches in the components' contracts with respect to their extra props.

React version: 16.13.1

## Steps To Reproduce

Have a look at the Codepen example: https://codepen.io/nicolerauch/pen/QWNLENx?editors=1010

## The current behavior
1. `FewerPropsClass` is not accepted (i.e. the line compiles) when it is passed to the `List` component.
1. `MorePropsFunctional` and `MorePropsClass` are accepted (i.e. there are no compile errors). If the surplus parameters are accessed in those components, this leads to runtime errors.

## The expected behavior
1. `FewerPropsClass` should be accepted when it is passed to the `List` component (i.e. the line should compile).
1. Neither `MorePropsFunctional` nor `MorePropsClass` should be accepted (i.e. there should be compile errors).

**Comment by bvaughn** [comment]
Am I reading this wrong, or is the bug being reported here a TypeScript issue? React has nothing to do with TypeScript. If you believe you've found a TypeScript bug, you should file it in the TypeScript repo: https://github.com/microsoft/TypeScript

**Comment by NicoleRauch** [comment]
Thanks for getting back to me!

My thinking was that the bug is in React's TypeScript typings (i.e. they being too strict or too lax), that's why I reported here. It did not even occur to me that this could be a genuine TypeScript issue...

But I just tried the behaviour with plain Typescript functions and classes, and it's identical (well, not surprising, I guess). So I will report this issue with the TypeScript team. (Sorry for the disturbance and for not checking this earlier!)
